src/c_kzg_4844.c

```diff
@@ -2369,344 +2369,151 @@ static C_KZG_RET ifft_fr(
 
-/**
- * Copy polynomial and set remaining fields to zero.
- *
- * @param[out]  out     The output polynomial with padded zeros
- * @param[out]  out_len The length of the output polynomial
- * @param[in]   in      The input polynomial to be copied
- * @param[in]   in_len  The length of the input polynomial
- */
-static C_KZG_RET pad_p(
-    fr_t *out, size_t out_len, const fr_t *in, size_t in_len
-) {
-    /* Ensure out is big enough */
-    if (out_len < in_len) {
-        return C_KZG_BADARGS;
-    }
+        poly[i] = neg_root;
+        blst_fr_add(&poly[i], &poly[i], &poly[i - 1]);
 
-    /* Copy polynomial fields */
-    for (size_t i = 0; i < in_len; i++) {
-        out[i] = in[i];
+        for (size_t j = i - 1; j > 0; j--) {
+            blst_fr_mul(&poly[j], &poly[j], &neg_root);
+            blst_fr_add(&poly[j], &poly[j], &poly[j - 1]);
+        }
+        blst_fr_mul(&poly[0], &poly[0], &neg_root);
     }
 
-    /* Set remaining fields to zero */
-    for (size_t i = in_len; i < out_len; i++) {
-        out[i] = FR_ZERO;
-    }
+    poly[roots_len] = FR_ONE;
+    *poly_len = roots_len + 1;
 
     return C_KZG_OK;
 }
 
 /**
- * Calculate the product of the input polynomials via convolution.
+ * Computes the minimal polynomial that evaluates to zero at equally spaced
+ * chosen roots of unity in the domain of size `FIELD_ELEMENTS_PER_BLOB`.
+ *
+ * The roots of unity are chosen based on the missing cell indices. If the i'th
+ * cell is missing, then the i'th root of unity from `expanded_roots_of_unity`
+ * will be zero on the polynomial computed, along with every
+ * `CELLS_PER_EXT_BLOB` spaced root of unity in the domain.
  *
- * @param[out]  out             Polynomial with len_out fields allocated
- * @param[in]   len_out         Domain evaluation length, a power of two
- * @param[in]   scratch         Scratch space, at least 3x len_out
- * @param[in]   len_scratch     Scratch size, at least 3x len_out
- * @param[in]   partials        Array of polys to be multiplied together
- * @param[in]   partial_count   The number of polys to be multiplied together
- * @param[in]   s               The trusted setup
+ * @param[in,out]   vanishing_poly          The vanishing polynomial
+ * @param[in]       missing_cell_indices    The array of missing cell indices
+ * @param[in]       len_missing_cells       The number of missing cell indices
+ * @param[in]       s                       The trusted setup
  *
- * @remark This will pad the polynomials, perform FFTs, point-wise multiply the
- *         results together, and apply an inverse FFT to the result.
+ * @remark When all of the cells are missing, this algorithm has
+ * an edge case. We return C_KZG_BADARGS in that case.
+ * @remark When none of the cells are missing, recovery
+ * is trivial. We expect the caller to handle this case,
+ * and return C_KZG_BADARGS if not.
+ * @remark `missing_cell_indices` are assumed to be less than
+ * `CELLS_PER_EXT_BLOB`.
```

Comment:
```suggestion
 * `CELLS_PER_EXT_BLOB` and at least one.
```

Reply:
The previous remark should be good enough.